File: internal/actor/actor.go
// Package actor classifies a scheduled-run actor login by account durability —
// the signal that decides whether a cron needs re-homing.
//
//	bot / service  -> durable automation; leave it.
//	deprovisioned  -> EMU-anonymized (former employee); runs have STOPPED.
//	human          -> a personal "*_EMU" handle; dies on departure.
//	external       -> a live non-EMU account (mirrored upstream); unowned.
//	none           -> no scheduled run on record.
package actor

import (
	"regexp"
	"strings"
	"sync"
)

// emuSuffix is the GitHub Enterprise Managed User handle suffix (e.g. "_acme").
// Personal EMU logins end with it; set it for your enterprise via SetEMUSuffix.
var (
	mu        sync.RWMutex
	emuSuffix = "_EMU"
	anonRE    = compileAnon(emuSuffix)
)

func compileAnon(suffix string) *regexp.Regexp {
	return regexp.MustCompile(`^[0-9a-f]{20,}` + regexp.QuoteMeta(suffix) + `$`)
}

// SetEMUSuffix overrides the enterprise EMU handle suffix used to tell a personal
// (human) account from an external one.
func SetEMUSuffix(suffix string) {
	mu.Lock()
	defer mu.Unlock()
	emuSuffix = suffix
	anonRE = compileAnon(suffix)
}

// EMUSuffix returns the currently configured EMU suffix.
func EMUSuffix() string {
	mu.RLock()
	defer mu.RUnlock()
	return emuSuffix
}

// Non-App bot actors seen firing crons (GitHub's web merge bot).
var bots = map[string]bool{"web-flow": true}

// ClassOrder ranks classes most-fragile first (handy for risk sorting).
var ClassOrder = map[string]int{
	"deprovisioned": 0, "none": 1, "external": 2,
	"human": 3, "service": 4, "bot": 5,
}

// classes whose crons should be moved onto a durable account.
var needsRehome = map[string]bool{
	"deprovisioned": true, "human": true, "external": true,
}

var disposition = map[string]string{
	"bot":           "leave (bot-owned)",
	"service":       "leave (svc bot-owned)",
	"deprovisioned": "URGENT re-home (actor deprovisioned)",
	"human":         "re-home (personal account)",
	"external":      "re-home (external account)",
	"none":          "inert (never fired here)",
}

// IsBot reports whether a login is a GitHub App / known bot actor.
func IsBot(login string) bool {
	return login != "" && (strings.HasSuffix(login, "[bot]") || bots[login])
}

// IsService reports whether a login is a svc-* automation account.
func IsService(login string) bool {
	return login != "" && (strings.HasPrefix(login, "svc-") || strings.HasPrefix(login, "svc_"))
}

// Class buckets a scheduled-run actor login by account durability.
func Class(login string) string {
	mu.RLock()
	defer mu.RUnlock()
	switch {
	case login == "":
		return "none"
	case IsBot(login):
		return "bot"
	case IsService(login):
		return "service"
	case anonRE.MatchString(login):
		return "deprovisioned"
	case strings.HasSuffix(login, emuSuffix):
		return "human"
	default:
		return "external"
	}
}

// NeedsRehome reports whether a cron fired by an actor of this class should be
// re-homed onto a durable account.
func NeedsRehome(class string) bool { return needsRehome[class] }

// Disposition returns the recommended action text for a cron given its
// run-actor class.
func Disposition(class string) string { return disposition[class] }

File: lib/actor.py
"""Scheduled-run actor classification by account durability.

The actor of a scheduled run (last_runs.json `actor`) is the ground truth for
who fires a cron. Its durability decides whether the cron needs re-homing:

  bot / service  -> durable automation; leave it.
  deprovisioned  -> EMU-anonymized (former employee); runs have STOPPED.
  human          -> a personal `*_LinkedIn` handle; dies on departure.
  external       -> a live non-LinkedIn account (mirrored upstream); unowned.
  none           -> no scheduled run on record.

Mirrors scripts/cron_owner_burndown.py so the two stay consistent.
"""

import re

# An EMU handle anonymizes to a long hex hash + `_LinkedIn` on deprovision.
ANON_RE = re.compile(r"^[0-9a-f]{20,}_LinkedIn$")

# Non-App bot actors seen firing crons.
BOTS = {"li-auto-merge", "web-flow"}

# Most-fragile first (handy for risk sorting).
ACTOR_CLASS_ORDER = {"deprovisioned": 0, "none": 1, "external": 2,
                     "human": 3, "service": 4, "bot": 5}

# Classes whose crons should be moved onto a durable account.
_NEEDS_REHOME = {"deprovisioned", "human", "external"}

DISPOSITION = {
    "bot": "leave (bot-owned)",
    "service": "leave (svc bot-owned)",
    "deprovisioned": "URGENT re-home (actor deprovisioned)",
    "human": "re-home (personal account)",
    "external": "re-home (external account)",
    "none": "inert (never fired here)",
}


def is_bot(login):
    return bool(login) and (login.endswith("[bot]") or login in BOTS)


def is_service(login):
    return bool(login) and (login.startswith("svc-") or login.startswith("svc_"))


def actor_class(login):
    """Bucket a scheduled-run actor login by account durability."""
    if not login:
        return "none"
    if is_bot(login):
        return "bot"
    if is_service(login):
        return "service"
    if ANON_RE.match(login):
        return "deprovisioned"
    if login.endswith("_LinkedIn"):
        return "human"
    return "external"


def needs_rehome(cls):
    """True if a cron fired by an actor of this class should be re-homed."""
    return cls in _NEEDS_REHOME


def disposition(cls):
    """Recommended action text for a cron given its run-actor class."""
    return DISPOSITION[cls]
